doc/index.md

```diff
@@ -291,6 +291,101 @@ sentinel. That is only necessary if the underlying range's sentinel type models
 BidirectionalIterator. That's a finer point that you shouldn't worry about right
 now.)*
 
+## Create Custom Iterators
+
+Here is example of Range v3 compatible RandomAcess proxy iterator.
+Iterator return std::pair key/value (aka zip).
+
+~~~~~~~{.cpp}
+    using KeyIter   = typename std::vector<Key>::iterator;
+    using ValueIter = typename std::vector<Value>::iterator;
+
+    struct cursor {
+
+        // basic_iterator derives from "mixin", if present, so it can be used
+        // to inject things into the public interface of the iterator
+        struct mixin : ranges::basic_mixin<cursor>
+        {
+            using ranges::basic_mixin<cursor>::basic_mixin;
+
+            // It is necessary to expose constructor in this way
+            mixin(KeyIter key_iterator, ValueIter value_iterator)
+                : mixin{ cursor(key_iterator, value_iterator) }
+            {}
+
+            KeyIter key_iterator() {
+                return this->get().key_iterator;
+            }
+            ValueIter value_iterator() {
+                return this->get().value_iterator;
+            }
+        };
+
+        // This is for dereference operator.
+        using cursor_value_t = std::pair<Key, Value&>;
+        cursor_value_t read() const {
+            return {*key_iterator, *value_iterator};
+        }
```

Comment:
Your value type should not have reference semantics. `std::pair<Key, Value>`. Then `read()` should return something like `ranges::common_pair<Key&, Value&>` that makes the iterator satisfy the syntactic requirements of the `Readable` concept.

Proxies are hard to get right. I hope for C++?? we get the conversion operators in `tuple` and `pair` to make this all Just Work.

Reply:
> Your value type should not have reference semantics

I see that `reference_t` is `read()`'s return type.
What is `value_type_t` for this iterator/cursor? Does `basic_iterator` infer it somehow from `cursor`? Should/could it be specified directly?

---

>  I hope for C++?? we get the conversion operators in tuple and pair to make this all Just Work.

What about custom types, like:
```
struct KeyValue{
     Key& key;
     Value& value;
};
```

Reply:
The problem of defining a proxy iterator is hard. I suggest you read [my blog post](http://ericniebler.com/2015/02/13/iterators-plus-plus-part-2/) on the subject, and then try to implement a proxy iterator _and test that it satisfies range-v3's `InputIterator` concept_.

Reply:
> I suggest you read my blog post on the subject

Actually I already read it... First two times when I read it, I just couldn't get why `value_type` is needed at all. Then I saw this https://stackoverflow.com/a/33115699...

BTW, there is `view::transform`, which can return `KeyValue` struct, or pair of references ... Maybe there is need to note, that such things are not valid... _Or they are?_

> and test that it satisfies range-v3's InputIterator concept.

It already satisfies it: https://wandbox.org/permlink/1humI5P3NagLS44W
Maybe, you mean that there should be also compare operators between them KeyValue and KeyValueRef?